python/lsst/pipe/base/graph/_implDetails.py

```diff
@@ -0,0 +1,85 @@
+# This file is part of pipe_base.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+from __future__ import annotations
+
+__all__ = ("_DatasetTracker", "_OutputSet", "_datasetDictToEdgeIterator")
+
+from dataclasses import dataclass, field
+from typing import (Generic, Mapping, Optional, Set, TypeVar, Generator, Tuple, NewType, Union)
+
+# Generic type parameters
+_T = TypeVar("_T")
+_U = TypeVar("_U")
+_S = TypeVar("_S")
+
+# NewTypes
+DSTypeName = NewType("DSTypeName", str)
+
+
+class _OutputSet(set, Generic[_T]):
+    """Simple set wrapper that provides a "first" method
+    to sets, that gets the first element of a set as if
+    it was being iterated over.
+
+    This is an internal implementation detail
+    """
+    def add(self, element: _T):
+        super().add(element)
+
+    @property
+    def first(self) -> Optional[_T]:
+        if len(self) == 0:
+            return None
+        return next(iter(self))
```

Comment:
I think this would be a lot better as free function; you don't want to have to copy some external `set` (or even convert some other `set`-like object into one of these) just to be able to use this method.

Reply:
Dropped in a refactor